src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,271 @@
+            private static readonly long PingIntervalInTicks =(long)(PingIntervalInSeconds * Stopwatch.Frequency);
+
+            private State _state;
+            private long _pingSentTimestamp;
+            private long _pingCounter;
+            private int _initialBurst;
+            private long _minRtt;
+
+            public TimeSpan MinRtt => new TimeSpan(_minRtt);
+
+            public static RttEstimator Create()
+            {
+                RttEstimator e = default;
+                e._state = GlobalHttpSettings.SocketsHttpHandler.DisableDynamicHttp2WindowSizing ? State.Disabled : State.Init;
+                e._initialBurst = InitialBurstCount;
+                return e;
+            }
+
+            internal void OnInitialSettingsSent()
+            {
+                if (_state == State.Disabled) return;
+                _pingSentTimestamp = Stopwatch.GetTimestamp();
+            }
+
+            internal void OnInitialSettingsAckReceived(Http2Connection connection)
+            {
+                if (_state == State.Disabled) return;
+                RefreshRtt(connection);
+                _state = State.Waiting;
+            }
+
+            internal void OnDataOrHeadersReceived(Http2Connection connection)
+            {
+                if (_state != State.Waiting) return;
+
+                long now = Stopwatch.GetTimestamp();
+                bool initial = _initialBurst > 0;
+                if (initial || now - _pingSentTimestamp > PingIntervalInTicks)
+                {
+                    if (initial) _initialBurst--;
+
+                    // Send a PING
+                    _pingCounter--;
+                    connection.LogExceptions(connection.SendPingAsync(_pingCounter, isAck: false));
+                    _pingSentTimestamp = now;
+                    _state = State.PingSent;
+                }
+            }
+
+            internal void OnPingAckReceived(long payload, Http2Connection connection)
+            {
+                if (_state != State.PingSent && _state != State.TerminatingMayReceivePingAck)
+                {
+                    ThrowProtocolError();
+                }
+
+                if (_state == State.TerminatingMayReceivePingAck)
+                {
+                    _state = State.Disabled;
+                    return;
+                }
+
+                //RTT PINGs always carry negavie payload, positive values indicate a response to KeepAlive PING.
+                Debug.Assert(payload < 0);
+
+                if (_pingCounter != payload)
+                    ThrowProtocolError();
+
+                RefreshRtt(connection);
+                _state = State.Waiting;
+            }
+
+            internal void OnGoAwayReceived()
+            {
+                if (_state == State.PingSent)
+                {
+                    // We may still receive a PING ACK, but we should not send anymore PING:
+                    _state = State.TerminatingMayReceivePingAck;
+                }
```

Comment:
As far as I understand the RFC, as part of the graceful termination process of outstanding streams, the server is still allowed to react to PINGs after sending a GOAWAY, so I had to add this state. Correct me if I was wrong.

Reply:
Yes, the peer can definitely respond to PINGs after sending GOAWAY.